src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,838 @@
+        let (Some(split_idx), Some(offset_idx)) = get_split_offset_col_idx(&source_desc.columns)
+        else {
+            unreachable!("Partition and offset columns must be set.");
+        };
+
+        let mut boot_state = Vec::default();
+        if let Some(mutation) = barrier.mutation.as_ref() {
+            match mutation.as_ref() {
+                Mutation::Add(AddMutation { splits, .. })
+                | Mutation::Update(UpdateMutation {
+                    actor_splits: splits,
+                    ..
+                }) => {
+                    if let Some(splits) = splits.get(&self.actor_ctx.id) {
+                        boot_state = splits.clone();
+                    }
+                }
+                _ => {}
+            }
+        }
+        self.backfill_state_store.init_epoch(barrier.epoch);
+
+        let mut backfill_states: BackfillStates = HashMap::new();
+        let mut unfinished_splits = Vec::new();
+        for mut split in boot_state {
+            let split_id = split.id();
+            let backfill_state = self
+                .backfill_state_store
+                .try_recover_from_state_store(&split_id)
+                .await?;
+            match backfill_state {
+                None => {
+                    backfill_states.insert(split_id, BackfillState::Backfilling(None));
+                    unfinished_splits.push(split);
+                }
+                Some(backfill_state) => {
+                    match backfill_state {
+                        BackfillState::Backfilling(ref offset) => {
+                            if let Some(offset) = offset {
+                                split.update_in_place(offset.clone())?;
+                            }
+                            unfinished_splits.push(split);
+                        }
+                        BackfillState::SourceCachingUp(_) | BackfillState::Finished => {}
+                    }
+                    backfill_states.insert(split_id, backfill_state);
+                }
+            }
+        }
+        tracing::debug!(?backfill_states, "source backfill started");
+
+        // Return the ownership of `stream_source_core` to the source executor.
+        self.stream_source_core = core;
+
+        let (source_chunk_reader, abort_handles) = self
+            .build_stream_source_reader(&source_desc, unfinished_splits.clone())
+            .instrument_await("source_build_reader")
+            .await?;
+
+        fn select_strategy(_: &mut ()) -> PollNext {
+            futures::stream::PollNext::Left
+        }
+        // XXX:
+        // - What's the best poll strategy? We should prefer backfill, but also consider barrier from input.
+        // - Should we also add a barrier stream for backfill executor?
+        let mut backfill_stream = select_with_strategy(
+            input.by_ref().map(Either::Left),
+            source_chunk_reader.map(Either::Right),
+            select_strategy,
+        );
```

Comment:
I think this is the most concerning part. Help wanted

Reply:
After some thoughts, I feel that maybe "preferring upstream" is an acceptable strategy. - When the upstream source's workload is high (i.e., Kafka has new incoming data), it just makes backfilling slower. For chunks from upstream, they are simply dropped, so there's no much overhead. So possibly this can also affect other running jobs less. 
- When the upstream Source's becomes less busy, SourceBackfill can begin to catch up. 

But currently in my local stress test, I found that current implementation still has barrier stuck issue. I'm suspecting it's because I created one Kafka consumer for each partition and they used too much memory. Still investigating...

Reply:
Just confirmed that if I use one rdkafka consumer for all partitions (like SourceExecutor does), the barrier doesn't stuck any more. TBO I still can't understand the root cause, not sure if it's related with "blocking calls" issue https://risingwave-labs.slack.com/archives/C03A2PSS8KU/p1709102606728729?thread_ts=1709101696.046829&cid=C03A2PSS8KU

Left side is 1 consumer for N partitions; Right side is N for N.
<img width="1653" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/37948597/6f91222b-e325-45e1-a15b-7f6c48047f48">

I used the N for N way for the "abort early" optimization. I guess this optimization isn't that useful if the partitions are balanced, and it's fine to read a few more unneeded data from finished partitions.